content/en/apps/reference/extension-libs.md

```diff
@@ -0,0 +1,89 @@
+---
+title: "extension-libs/"
+linkTitle: "extension-libs/"
+weight: 5
+description: >
+  Used for providing custom scripts for execution in CHT apps
+keywords: extension libs extensibility
+---
+
+_Introduced in v4.2.0_
+
+## Introduction
+
+Extension libraries are blocks of code that are cached with the CHT web application giving app developers a powerful tool to extend the CHT. This is an advanced feature and requires an app developer with some software development experience.
+
+An example of a use for this feature is to provide a function to calculate a risk score based on a machine learning model. The function can then be called passing in values from app forms and return the result to be stored with the report.
+
+### Library
+
+The first step is to create the js file to return the function that will be called by the web application. Create a new file using this template:
+
+```js
+return function(/* parameters */) {
+  return result;
+}
+```
+
+Now populate the function as needed. For complex functions it is recommended to use third party libraries and use a bundler (eg: [webpack](https://webpack.js.org/)) to make it easy to build a single file. For anything non-trivial, it's recommended to use development best practices such as linting, unit tests, and minification.
+
+#### xpath functions
+
+To call the function from within a form the parameters and return value will need to have a very specific structure to work with Enketo xforms.
+
+```json
+{
+  "t": <type>,
+  "v": <value>
+}
+```
+
+Where the type is one of "bool", "num", "str", "date", or "arr". For example, to calculate the average of two inputs, you would use:
+
+```js
+const getValue = function(obj) {
+  let val;
+  if (obj.t === 'arr') {
+    val = obj.v && obj.v.length && obj.v[0];
+  } else {
+    val = obj.v;
+  }
+  if (!val) {
+    return 0;
+  }
+  const parsed = parseInt(val.textContent);
+  return isNaN(parsed) ? 0 : parsed;
+};
+
+return function(first, second) {
+  const average = (getValue(first) + getValue(second)) / 2;
+  return {
+    t: 'num',
+    v: average
+  };
+}
+```
+
+### Uploading to the CHT
+
+Create a folder within the project configuration to contain the libraries, for example:
+
+```
+./extension-libs
+    average.js
+    calculate-risk-score.js
```

Comment:
Random question, but as I looked at this, the first thing that popped into my head is what code-reuse would look like in this process.  Say I have a `utils.js` file that contains code that is common between `average.js` and `calculate-risk-score.js`.  Even if I am using webpack to create average/calculate-risk-score so that the `utils.js` code is included in those files, does that mean that the code is going to be duplicated in the extension attachments (and ultimately duplicated in webapp)?  With separate files for each function are we forcing all re-used code to be duplicated?

If so, I wonder if it would be better to just require that all extension functions come from a single file.  (Alternatively, would it make more sense to have app devs include all their code in the `/extension-libs` folder and then cht-conf can handle webpacking everything before attaching it?)

Reply:
Part of the idea is to keep functions separate so if you change one then users only need to download the new one rather than the whole bundle again. An example for this is including the momentjs library so you can use it in the tasks config.

You're right though, if you also want to call momentjs from your risk score calculator then you'd introduce duplication. How do you feel about making the script api available to the extension libs, so you can get a lib fn from inside a lib fn?